### PR TITLE
add --ignore-unknown-parameters flag to new-app and process

### LINKED ISSUE #12909: 'oc process' should be able to ignore invalid parameter names via a flag
`oc process` in oc version 3.3.1.7 throws error and stops processing template when it encounters param in `-v` that is not defined in template. Why is that? It should be treated as a warning and processing should continue.

In fact the oc version 3.2.0.44 behaved exactly like that - it displayed warning but did not stop processing.

##### Version
3.3.1.7

##### Steps To Reproduce
1. process template file with `oc process -f template.yml -v VAR=value`, whereas `VAR` is not defined in template

##### Current Result
command fails due to unrecognized param with message: `error: unknown parameter name "VAR"`
##### Expected Result
should display warning (like it was in version 3.2.0.44, but should not fail.

### PR DESCRIPTION
**Pull request by matthyx**
'oc process' should be able to ignore invalid parameter names via a flag 
fixes #12909

### CODE REVIEW

**Comment by bparees** [comment]
[test]

**Comment by mfojtik** [comment]
@bparees don't we have 'required' parameter for this?

**Comment by bparees** [comment]
> @bparees don't we have 'required' parameter for this?

no that's different.  required means "you must supply a non-empty string value for this parameter".

this PR is to allow users to do something like:
oc proess -f mytemplate -p some_param_name_that_doesnt_exist=foo  --ignore-invalid-parameters

and not get an error.

Your confusion on the subject means the flag name is probably not great since in this case we mean "invalid" as in "does not exist", not "invalid" as in "not a valid value for the parameter".

@matthyx can we try a different flag name?  maybe "--skip-absent-parameters"?

**Comment by bparees** [comment]
[test]

**Comment by pweil-** [comment]
@openshift/cli-review

**Comment by matthyx** [comment]
@bparees I have the change with the new parameter name... should I push now or I wait until I have tests ready as well?

**Comment by bparees** [comment]
@matthyx might as well add the tests first.

**Comment by fabianofranz** [comment]
Some ideas

```
--ignore-unknown-parameters
--allow-unknown-parameters
--skip-undeclared-parameters
--skip-param-validation
```

**Comment by bparees** [comment]
`--ignore-unknown-parameters` works for me.

**Comment by matthyx** [comment]
@bparees done with tests added and squashed

File: pkg/cmd/cli/cmd/newapp.go [hunk 1]
```diff
@@ -237,6 +237,7 @@ func NewCmdNewApplication(name, baseName string, f *clientcmd.Factory, in io.Rea
 	cmd.Flags().StringVar(&config.Name, "name", "", "Set name to use for generated application artifacts")
 	cmd.Flags().Var(&config.Strategy, "strategy", "Specify the build strategy to use if you don't want to detect (docker|pipeline|source).")
 	cmd.Flags().StringP("labels", "l", "", "Label to set in all resources for this application.")
+	cmd.Flags().BoolVar(&config.IgnoreUnknownParameters, "ignore-unknown-parameters", false, "If true, will not stop processing the template when a parameter is not valid.")
```
Comment (bparees) [inline pkg/cmd/cli/cmd/newapp.go]: help text should be "If true, will not stop processing if a provided parameter does not exist in the template."

File: pkg/cmd/cli/cmd/process.go [hunk 2]
```diff
@@ -92,6 +92,7 @@ func NewCmdProcess(fullName string, f *clientcmd.Factory, in io.Reader, out, err
 	cmd.Flags().StringArrayVarP(params, "param", "p", nil, "Specify a key-value pair (eg. -p FOO=BAR) to set/override a parameter value in the template.")
 	cmd.Flags().StringArray("param-file", []string{}, "File containing template parameter values to set/override in the template.")
 	cmd.MarkFlagFilename("param-file")
+	cmd.Flags().Bool("ignore-unknown-parameters", false, "If true, will not stop processing the template when a parameter is not valid.")
```
Comment (bparees) [inline pkg/cmd/cli/cmd/process.go]: same help text change here.

**Comment by bparees** [comment]
[test]

**Comment by matthyx** [comment]
@bparees I did the change and squashed... it seems github would have preferred I did on the web interface instead :-/

**Comment by bparees** [comment]
[merge][severity:lowrisk]

**Comment by bparees** [comment]
thanks @matthyx!

**Comment by bparees** [comment]
flake https://github.com/openshift/origin/issues/15333
flake https://github.com/openshift/origin/issues/13463
flake https://github.com/openshift/origin/issues/14897

[merge]

**Comment by bparees** [comment]
[merge]

**Comment by gunjansharma18** [comment]
In which version is this fixed ? I am using OCP 3.5 and still able to reproduce thos error ?

**Comment by bparees** [comment]
3.5 is pretty old (3.9 is current and 3.10 is about to release).  the fix appears to be in v3.7

**Pull request merged**